### ISSUE TITLE
Istio gateways are hard coded and anti-pattern

### ISSUE THREAD

**Issue by adamkgray** [kubeflow/kfserving#1340]
![kfserving](https://user-images.githubusercontent.com/29312513/106587255-5caf1a80-6584-11eb-81c9-61115e65afc7.png)
/kind feature

**Describe the solution you'd like**

You should be able to specify any (or none!) ingress-gateway/cluster-local-gateway when using kfserving

Right now the values are hardcoded in `pkg/constants/constants.go`:
```
// Knative constants
const (
    KnativeLocalGateway   = "knative-serving/cluster-local-gateway"
    KnativeIngressGateway = "knative-serving/knative-ingress-gateway"
    VisibilityLabel       = "serving.knative.dev/visibility"
)

var (
    LocalGatewayHost = "cluster-local-gateway.istio-system.svc." + network.GetClusterDomainName()
)
```

It is a misnomer because in knative these are not strictly constants; you can use any gateway you want

You ought to be able to configure it like so:
```
// Knative constants
const (
    KnativeLocalGateway   = "<namespace>/<gateway name>"
    KnativeIngressGateway = "<namespace>/<gateway name>"
    VisibilityLabel       = "serving.knative.dev/visibility"
)

var (
    LocalGatewayHost = "<gateway/envoy name>.<namespace>.svc." + network.GetClusterDomainName()
)
```

Moreover when a InferenceService creates an istio VirtualService it sets the route from the host (example `sklearn-iris.default.example.com`) to `cluster-local-gateway.istio-system.svc.cluster.local` and sets the host header as `iris-predictor-default.default.svc.cluster.local`. In my understanding this is fundamentally an incorrect use of the cluster local gateway. 

The source says:
```
    # A cluster local gateway to allow pods outside of the mesh to access
    # Services and Routes not exposing through an ingress.  If the users
    # do have a service mesh setup, this isn't required and can be removed.
    #
    # An example use case is when users want to use Istio without any
    # sidecar injection (like Knative's istio-ci-no-mesh.yaml).  Since every pod
    # is outside of the service mesh in that case, a cluster-local  service
    # will need to be exposed to a cluster-local gateway to be accessible.
 ```

The cluster local gateway should be an entrance point for services which *do not* run in the mesh. If you are doing everything according to best practices you don't even need it. So a possible fix would be to route the traffic directly to the service, namely `iris-predictor-default.default.svc.cluster.local` and not through the cluster local gateway at all.

**Anything else you would like to add:**

TLDR Kfserving should let you BYO knative and istio with no assumptions. Moreover, by forcing the need for a cluster-local-gateway, this breaks potential for mesh-only setups.

Imho it is strange that kfserving by default wants to serve on a ingress rather than only in the mesh - it makes much more sense in a production setting to have models served only internally and let some other api do the dealing with callers.

**Comment by yuzisun** [comment]
@adamkgray Thanks for raising this issue! We are working on to make those constants configurable, to answer your questions.

```
The cluster local gateway should be an entrance point for services which do not run in the mesh. If you are doing everything according to best practices you don't even need it. So a possible fix would be to route the traffic directly to the service, namely iris-predictor-default.default.svc.cluster.local and not through the cluster local gateway at all.
```
I do not think this is true, `sklearn-iris-predictor-default.default.svc.cluster.local` is just an external name which points to cluster local gateway.
```
kubectl get svc sklearn-iris-predictor-default
NAME                              TYPE           CLUSTER-IP   EXTERNAL-IP                                            PORT(S)   AGE
sklearn-iris-predictor-default   ExternalName   <none>       cluster-local-gateway.istio-system.svc.cluster.local   <none>    21h
```
